language/move-binary-format/src/control_flow_graph.rs

```diff
@@ -73,32 +93,104 @@ impl VMControlFlowGraph {
         }
 
         // Create basic blocks
-        let mut cfg = VMControlFlowGraph { blocks: Map::new() };
+        let mut blocks = Map::new();
         let mut entry = 0;
+        let mut exit_to_entry = Map::new();
         for pc in 0..code.len() {
-            let co_pc: CodeOffset = pc as CodeOffset;
+            let co_pc = pc as CodeOffset;
 
             // Create a basic block
-            if VMControlFlowGraph::is_end_of_block(co_pc, code, &block_ids) {
+            if Self::is_end_of_block(co_pc, code, &block_ids) {
+                let exit = co_pc;
+                exit_to_entry.insert(exit, entry);
                 let successors = Bytecode::get_successors(co_pc, code);
-                let bb = BasicBlock {
-                    entry,
-                    exit: co_pc,
-                    successors,
-                };
-                cfg.blocks.insert(entry, bb);
+                let bb = BasicBlock { exit, successors };
+                blocks.insert(entry, bb);
                 entry = co_pc + 1;
             }
         }
+        let blocks = blocks;
+        assert_eq!(entry, code_len);
+
+        // Determine traversal order
+        // build a DAG subgraph (remove the loop back edges)
+        let dag: Map<BlockId, Set<BlockId>> = blocks
+            .iter()
+            .map(|(id, block)| {
+                let id = *id;
+                let non_loop_continue_successors = block
+                    .successors
+                    .iter()
+                    // remove the loop back edges
+                    .filter(|successor| **successor > id)
```

Comment:
Is this sufficient to check that it's a back edge? I would think in principal we could have a CFG like

```
B1
  goto B3;
B2:
  return
B3:
  goto B2;
```

I think this code will treat B3 -> B2 as a loop back edge, but it is not.

Reply:
That code would not be allowed in our current CFG set of checks (see language/move-bytecode-verifier/src/control_flow.rs). This would be counted as "jumping into the middle of a loop"

Reply:
Can this check now do something like `!is_back_edge(successor, id)` instead of `**successor > id`